src/freertos_drivers/arduino/ArduinoGpio.hxx

```diff
@@ -0,0 +1,268 @@
+/** \copyright
+ * Copyright (c) 2019, Mike Dunston
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are  permitted provided that the following conditions are met:
+ *
+ *  - Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  - Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * \file ArduinoGpio.hxx
+ *
+ * Helper declarations for using GPIO pins via the Arduino APIs.
+ *
+ * @author Mike Dunston
+ * @date 14 January 2019
+ */
+
+#ifndef _DRIVERS_ARDUINOGPIO_HXX_
+#define _DRIVERS_ARDUINOGPIO_HXX_
+
+#include "os/Gpio.hxx"
+#include "GpioWrapper.hxx"
+#include <Arduino.h>
+
+/// Defines a GPIO output pin. Writes to this structure will change the output
+/// level of the pin. Reads will return the pin's current level.
+///
+/// The pin is set to output at initialization time, with the level defined by
+/// `SAFE_VALUE'.
+///
+/// Do not use this class directly. Use @ref GPIO_PIN instead.
+template <int PIN_NUM>
+class ArduinoStaticGpio
+{
+public:
+    /// Sets pin to output.
+    static void set_output()
+    {
+        pinMode(PIN_NUM, OUTPUT);
+    }
+
+    /// Sets pin to input.
+    static void set_input()
+    {
+        pinMode(PIN_NUM, INPUT);
+    }
+
+    /// Turns on pullup.
+    static void set_pullup_on()
```

Comment:
pullup is enabled in two ways:
1) pinMode(pin, INPUT_PULLUP);
2) pinMode(pin, INPUT); digitalWrite(pin, HIGH);

I'm going to rework this slightly after discussion with @me-no-dev as there isn't a very good approach to use here for detecting the pin mode at runtime without storing the type/state in this class and it will quickly get out of sync if pinMode is used externally.

Reply:
I'll be rewriting this when I get back home Friday to use the IDF API instead of Arduino API as it offers a lot more flexibility and runtime detection of the current pinmode.

Reply:
Based on what you say it seems this code is correct now.

Reply:
Correct but confusing and limited a bit. I'll be looking at the IDF API as it exposes a few more options and is compatible with the Arduino API that others may end up using as well.